### ISSUE TITLE
Request: Hyperjump arrival/departure distance variables

### ISSUE THREAD

**Issue by Zitchas** [endless-sky/endless-sky#4334]
**TLDR summary: Just to be clear: This request specifies that the default values should be ZERO, which would mean that they have no impact on the current gameplay. Then these variables would be available for use by content creators (particularly plug-in makers) who *do* want this behaviour.**

**Is your feature request related to a problem? Please describe.**
1) During almost all combat missions I arrive much too close to the system center, and thus am already in close-range combat immediately upon exiting hyperjumps. 
2) Some people would like a more "Classic Escape Velocity (EV/EVO/EVN) feel to the game with mandatory minimum distance from the system center before one can hyperjump (and arrive from hyperspace)

**Related Issue Links**
Some conversation related to this:
https://github.com/endless-sky/endless-sky/issues/1094
https://github.com/endless-sky/endless-sky/issues/1635
https://github.com/endless-sky/endless-sky/issues/3518
https://github.com/endless-sky/endless-sky/issues/3661

A PR that made some steps towards it, but was never merged:
https://github.com/endless-sky/endless-sky/pull/3628

**Describe the solution you'd like**
A) Implementation of `hyperjump arrival radius` and a `hyperjump departure radius` variable, ideally as attributes of systems. These values would default to 0, and in their default configuration would have no impact on current gameplay mechanics of Endless Sky.
B) Implementation of `hyperjump arrival distance` and `hyperjump departure distance` attributes for ships and outfits. This is intended to allow certain ships to have different arrival or departure radii, and to allow outfits to change these values. 

All of these variables would use the normal system measurements to describe distances.

**Impact and Usage**
In vanilla Endless Sky, one immediate usage that I can see would be using these variables to create an outfit (say, the "Long Range Tactical Computer") that adds an amount to the hyperspace arrival radius. This would be immediately useful for ships like the Albatross (Sniper) variants that are built for longer range engagements. Variations on this might be very useful for various other configurations of NPC ships that prioritize longer-range engagements.  That being said, it is important to note that this would *not* be another hyperdrive, but rather it would be a small outfit that modifies the behavior of whatever drive the ship has installed, and would apply equally to hyperdrives and jump drives.

This would expand the tactical flexibility of the game, and provide opportunities for smart players to arrive into battles with a bit more time to adjust. This could significantly decrease the "jump and pray" aka save-scumming methods of dealing with certain large battles in the storyline.

This could also be used to change the feel of certain areas of the game. For instance, the Blackhole, the Ember Waste, and other unusual areas could have variable amounts of hyperspace interference near the centers of the systems which could be used to make these areas feel more dangerous and risky. (of course, in-game descriptions could be added so the player knows or finds out what is going on.)

Beyond Vanilla content, this change would also facilitate a variety of plugins, including:
- The EV/O/N total conversion plugins
- An "change ES to feel more like classic EVN" plug-in

**Other thoughts**
Over the years there's been plenty of interest in the community about implementing hyperjump distance restrictions for various reasons. Implementing it in this manner would put the tools in the hands of community members to use plug-ins to modify the gameplay to match their desires, while preserving the core gameplay as MZ mandated it should be. At the same time, it would allow for some interesting outfits and situations in-game that could add interest, tactical decision making, and atmosphere to various parts of the base game.

**Variations**
If it makes implementation easier, it would be quite reasonable to implement only half of this initially. For instance, implementing only the `hyperjump arrival radius` attribute for systems and the matching `hyperjump arrival distance` attribute for ships and outfits.

In terms of gameplay impacts and tactical uses described above, the arrival part of it would probably convey the largest benefit and utility. Given that the player and ships don't jump into the middle of the system anyway, I suspect the game mechanics for this are already present to some degree.

**Comment by Fzzr** [comment]
The limit to jumping out has been no'd before in #1094 as you linked, and I agree with that call. It would be frustrating to add yet another variable to the requirements for travelling around. For example, basic trading, which involves landing on a planet, menuing, and then launching again, would take substantially more time after each launch (at least on the way in you'll have some momentum, so it won't be as bad). The jump-out restriction would also have major implications for combat balance, as the relative ease right now of escaping from tight situations is a big part of early-game balance especially.

I definitely like the idea of the jump-in radius, though. It's really annoying to jump into a firefight and lose half my shields and sometimes some of my escorts before even turning around to engage. The balance implications here would be on the easy rather than hard side, which I suspect would be much easier to correct. This also doesn't require major new AI behavior, unlike the jump-out restriction.

**Comment by wsaul** [comment]
I desire this as an attribute attached to stellar objects

```	
	object
		sprite star/g5
		period 10
		exclusion 500
	object
		sprite planet/rock6
		distance 158.61
		period 31.9607
		exclusion 200
	object "New Boston"
		sprite planet/cloud6
		distance 513.86
		period 186.375
		exclusion 500
	object
		sprite planet/rock13-b
		distance 864.75
		period 406.87
		exclusion 300

**Comment by Zitchas** [comment]
I think whether or not it is attached to stellar objects or to the system as a whole depends on the coders.

I think that the arrival distance would work best based on the system center, thus avoidign the necessity for ES to calculate areas and orbits for everything and figure out how the distance interacts with variable areas of exclusion. 

It might work fairly well for as an exclusion for jumping out, though.

**Comment by Zitchas** [comment]
@Fzzr In regards to the jump-out distance, I would like to point out that I specified that the default would be 0 for both, with no impact. So implementing all aspects of this wouldn't have any change to most of the game. It would, however, enable people to use these limitations in plug-ins or special circumstances. 

For example, the black hole could be given a jump-out exclusion zone, as the gravitational distortion impacts the ability of hyperdrives to function. That would give the black hole a concrete in-game effect without having to mess around with the physics of making a gravitational well of some kind.

Or small pockets of exclusion zones could be put in the Ember Waste systems as a sort of  hazard to make the place feel that much less hospitable.

**Comment by Tadrix** [comment]
Arriving a bit farther away from the system centre might be reasonable, but not farther away from the planets (so it mostly applies only to uninhabited systems, unless jumping to the target planet's vicinity gets disabled). Might not be worth the trouble.

MZ has stated previously that he doesn't intend to use the EV mechanics of departing from systems, and tbh I can see why (as it was, it was rather clumsy).

Besides, it would be even more frustrating to new players (not only are you threatened with an RNG missile instakill, but you can't run away either). This would introduce massive problems and a need for yet another rebalance.

**Comment by Zitchas** [comment]
@Tadrix I think the case for why these mechanics (especially the jump departure one) shouldn't be implemented with non-zero values has been more than sufficiently made in previous requests. But as previously stated, this request isn't asking for that. **It is asking for them to be implemented with a default value of 0, meaning no noticeable change from the way things work currently.** 

Implemented in that way, since they are actually variables that can be set in the system properties and influenced by ship and outfit stats, this would open up these settings for use by content creators to use as they will. Basically, this request is all about opening up these values to content creators instead of keeping them locked somewhere in the source code. After all, we all arrive and depart from systems. While we depart from anywhere, there is clearly already some kind of rules determining where we arrive from. Exposing that value to modifications from within the data simply makes it into just one more attribute we can set instead of something hard-coded. 

Admittedly, the departure restriction doesn't appear to have any such rules, so it'd probably take more work to implement. So this request would make sense to implement partially, just doing the arrival part first, then the departure part later.

When someone happens to make a "Classic EV hyperjumping" plug-in, I don't think MZ would object to it anymore than he'd object to cheat plug-ins or plug-ins that give the player invincible ships. And we don't rebalance the game for the sake of plug-ins, either. So that's a non-issue.

**Comment by Tadrix** [comment]
A fair point (I've removed my downvote). It is not strictly necessary for the game, though, so good luck getting somebody to implement it and submit a PR.

**Comment by ChamEV** [comment]
I was looking at ways of making combat more engaging for the player, and one of the issues is all the ships just thrown on top of each other.  Adding an increased jump in distance that is conditional on if there are enemies in the system would help alleviate this, and when new enemies arrive to fight the player, they don't land right on top of them either.

**Comment by Pointedstick** [comment]
> I was looking at ways of making combat more engaging for the player, and one of the issues is all the ships just thrown on top of each other.

+1,000

Regardless of how this is fixed, I think it's a real problem.

**Comment by Zitchas** [comment]
From a story/cinematic point of view, huge fleet fights tend to be more dramatic when you have at least a few seconds for the realization to sink in that this is going to be a *big* fight. Much more impactful than just getting thrown into it and afterwards thinking "wow, that was big".

**Comment by SirStigII** [comment]
This would also make a cinematic missile exchange as well, which will add to the ambience.

**Comment by Zitchas** [comment]
Definitely! Makes missiles more useful (which should make the people who have complained about that happy), and makes those of us who have anti-missile ships in our fleet happy. (like me).

**Comment by petervdmeer** [comment]
Would it make sense to also have an arrival distance based on the amount of traffic already in the system? (Lots of ships: arrive further from the center)

**Comment by petervdmeer** [comment]
And (from a discussion with @beccabunny on Discord today) maybe also let the arrival distance for ships entering a system to go into a fight depend on the maximum range of their weapons?

So if a longe-range missile ship enters the system, then let it arrive just outside the range of the long-range missiles?

**Comment by SirStigII** [comment]
Just to add to the discussion, the main dev from EV:Override is trying to figure out the best way to save it now that ambrosia seems to be shutting down, and raised the arrival/departure in ES as a possible issue to porting it over.

**Comment by Fzzr** [comment]
> Just to add to the discussion, the main dev from EV:Override is trying to figure out the best way to save it now that ambrosia seems to be shutting down, and raised the arrival/departure in ES as a possible issue to porting it over.

Woah, where's that discussion happening?

**Comment by beccabunny** [comment]
Another idea was the possibility to add the possibility to add a multiplier to drives as an attribute, for an example to have a Remnant hyperdrive that, to scout and use long range weapons more effectively, arrives further from the center. The multiplier could be:
"Jump distance" 1 so the drive makes the ship arrive at the defined arrival distance variable (probably, 1250),
"Jump distance" 4 to arrive, in the example above, at a distance of 5000 
If no "Jump distance" attribute is specified, it is default to 0, meaning that the arrival distance variable won't have any effect even if specified to a system




> Woah, where's that discussion happening?

https://www.reddit.com/r/evnova/comments/cohegk/evo_keeping_ev_override_in_print/

**Comment by Fzzr** [comment]
> https://www.reddit.com/r/evnova/comments/cohegk/evo_keeping_ev_override_in_print/

Thanks, I'll check it out.

**Comment by Amazinite** [comment]
I'd certainly be interested in having whatever functionality that is needed for a true EV port be added to the ES engine.

**Comment by Fzzr** [comment]
> I'd certainly be interested in having whatever functionality that is needed for a true EV port be added to the ES engine.

I am immediately fine with hyperspace borders if that's what it takes.

**Comment by Zitchas** [comment]
@petervdmeer As a base mechanic, I don't think it makes sense for ships to vary their arrival distance based on the number of ships currently in-system (or whether or not they are hostile, for that matter). Mainly due to the fact that our current hyperdrive lore makes no mention of sensors or the gravitational impact of the presence of other ships.

On the other hand, this would make for an awesome specialty effect for a unique drive for some faction if it was tied into some kind of lore about them having long range scanners or developing new tech related to scanners and drives.

**Comment by petervdmeer** [comment]
Tehhowch made a good point on Discord that we need to identify all desired jump in/jump out behavior modifiers that modders will want to make sure the keywords we would use don't conflict with other behavior. (We will however not implement them all in 1 go.)

Here is a list/proposal for all the relevant keywords I could come up with:

`arrival <distance> ["link"] ["jump"]`
System property and/or planet property. The <distance> indicates the minimum amount of distance (from respectively the system center or the planets center) at which ships arrive when entering the system.
If the property is given on multiple places, then the maximal distance is used.
If link keyword is given, then the arrival distance only applies to incomming hyperlinks.
If jump keyword is given, then the arrival distance only applies to incomming jumps.

`departure <min> [<max>] ["link"] ["jump"]`
	`angle <nr>`
System property and/or planet property. The <min> indicates the minimum distance that a ships needs to be away from respectively the system center or planet center before the ship can start a jump.
If a maxium is given, then a ship needs to be no more than distance away from the centers to be able to leave.
If an angle is given, then the ship needs to be at the proper distance, but also in the proper direction of leaving (within the range given by angle).
If multiple properties are given, then all departure constraints need to be fullfilled before a ship can make the jump.
Content creators need to be aware of blocking jumps if constraints are conflicting, they might also acutally want to use this, for example by having a system where jumps are only possible once every four days (due to planet alignments).

`arrival <min>`
`departure <min>`
Ships properties. <min> gives the minimum arrival and departure distances for ships. If a system or planet also has minimum arrival and departure distances then the maximum of all distances will be used for the actual distance.
I considered to also allow ship properties to override (or modify by a multiplier) the planet and system properties, but I expect that we will not want to do that in the end. If some content creator has a good reason not to allow jumps within range of a certain planet, then I expect the story immersion to be more imporant than being able to override the ranges.
(And if overriding by explicit number or modifier is very much desired, then it can just be added using this keyword anyway.)

(System property) `combat visibility`
Allow ships to arrive at the most appropriate distance for combat (based on their personality).
When this is false, then all ships will arrive as if the system has no combat.
When this is true, then fleets with "heroic" and other military personalities could arrive just beyond their maximum weapons range (allowing the long-range weapons to be used before short-range weapons are used).
When this is true, then unarmed carriers could arrive outside weapons ranges whenever possible.
When this is true, then fleets with merchant type of personalities could arrive at a large distance from the combat area (trying to avoid combat), based on their personalities.

(Ship property) `combat detection <value> ["link"] ["jump"]`
The ability (strength) of a ship to detect combat and combatants (in general or for hyperlink or jump only). Related to the `combat visibility` system property.

(Planet property) `arrive focusable`
Allow automatically aiming for this planet when arriving in the system.

If you see more desired possible modifiers, then please post them.

**Comment by Zitchas** [comment]
That's a good list. I like the variety and options those present. Honestly, I'd be happy just with the arrival and departure min distances, but the res (especially the angle one) has some good possibilities too.

**Comment by Amazinite** [comment]
Oops. Got auto closed on a PR that didn't fully resolve this.
